src/Components/RoleToolbar.js

```diff
@@ -0,0 +1,227 @@
+  anySelected,
+  clearFiltersButton,
+  perPage,
+  nameFilter,
+  setNameFilter,
+  AccessRequestsPagination,
+  applications,
+}) => {
+    const [isBulkSelectOpen, setIsBulkSelectOpen] = React.useState(false);
+    // Filtering
+    const [isDropdownOpen, setIsDropdownOpen] = React.useState(false);
+    const [filterColumn, setFilterColumn] = React.useState(columns[0]);
+    const [isSelectOpen, setIsSelectOpen] = React.useState(false);
+    const selectLabelId = 'filter-application';
+    const selectPlaceholder = 'Filter by application';
+    const hasFilters = appSelections.length > 0 || nameFilter;
+
+    const onSelectAll = (_ev, isSelected) => {
+      if (isSelected) {
+          setSelectedRoles(filteredRows.map((row) => row.display_name));
+      } else {
+          setSelectedRoles([]);
+      }
+    };
+
+  return (
+  <React.Fragment>
+    <Toolbar id="access-requests-roles-table-toolbar">
+      <ToolbarContent>
+        <ToolbarItem>
+          <Dropdown
+            onSelect={() => setIsBulkSelectOpen(!isBulkSelectOpen)}
+            position="left"
+            toggle={
+              <DropdownToggle
+                splitButtonItems={[
+                  <DropdownToggleCheckbox
+                    key="a"
+                    id="example-checkbox-2"
+                    aria-label={anySelected ? 'Deselect all' : 'Select all'}
+                    isChecked={isChecked}
+                    onClick={() => onSelectAll(null, !anySelected)}
+                  />,
+                ]}
+                onToggle={(isOpen) => setIsBulkSelectOpen(isOpen)}
+                isDisabled={rows.length === 0}
+              >
+                {selectedRoles.length !== 0 && (
+                  <React.Fragment>
+                    {selectedRoles.length} selected
+                  </React.Fragment>
+                )}
+              </DropdownToggle>
+            }
+            isOpen={isBulkSelectOpen}
+            dropdownItems={[
+              <DropdownItem key="0" onClick={() => onSelectAll(null, false)}>
+                Select none (0 items)
+              </DropdownItem>,
+              <DropdownItem
+                key="1"
+                onClick={() =>
+                  setSelectedRoles(
+                    selectedRoles.concat(pagedRows.map((r) => r.display_name))
+                  )
+                }
+              >
+                Select page ({Math.min(pagedRows.length, perPage)} items)
+              </DropdownItem>,
+              <DropdownItem key="2" onClick={() => onSelectAll(null, true)}>
+                Select all ({filteredRows.length} items)
+              </DropdownItem>,
+            ]}
+          />
+        </ToolbarItem>
+        <ToolbarItem>
+          <InputGroup>
+            <Dropdown
+              isOpen={isDropdownOpen}
+              onSelect={(ev) => {
+                setIsDropdownOpen(false);
+                setFilterColumn(ev.target.value);
+                setIsSelectOpen(false);
```

Comment:
React 17 (version we currently use) has no state update batching. Meaning that each of these setState calls will force a re-render.

You can get around this by consolidatng the multiple state variables into `useReducer` instead a do a batch updates.

So for this components, I would just do that and replace all the `useState` hooks with one `useReducer`. https://reactjs.org/docs/hooks-reference.html#usereducer

Reply:
I've figured out how to use the userReducer in place of the useState. Let me know if its a good implementation of it.